### PR TITLE
`Learning paths`: Redesign competency graph

### PR DESCRIPTION
**Pull request by JohannesWt**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Similar to the student learning path UI the competency graph for students should be redesigned. This PR is part of the redesign of the complete student learning path UI. Therefore it will be reviewed and tested before getting merged into the base PR #8733. (which then will be merged into develop)

### Description

Students should have the ability to quickly see how the competencies they achieve in the learning path are roughly connected to each other.

### Steps for Testing

Prerequisites:
- 1 Students
- at least 2 competencies connected to each other (assumes, extends or similar)
- at least 2 lecture units (or exercises) each connected to a competency

1. Log in to Artemis
2. Navigate to Learning paths student view
3. Open the navigation overview
4. Open the competency dependency graph

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

#### Before
image

#### After
image

### CODE REVIEW

File: src/main/webapp/app/entities/competency/learning-path.model.ts [hunk 1]
```diff
@@ -44,6 +44,28 @@ export interface LearningPathNavigationOverviewDto {
...
+}
+
+export interface CompetencyGraphEdgeDto {
+    id: string;
+    source: string;
+    target: string;
+    relationType: CompetencyRelationType;
+}
+
+export interface CompetencyGraphDto {
```
Comment (pzdr7) [inline src/main/webapp/app/entities/competency/learning-path.model.ts:64]: Regarding DTOs, I think we have the convention that we spell it `DTO` and not `Dto`

File: src/main/webapp/app/course/learning-paths/components/competency-node/competency-node.component.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    private readonly element: ElementRef = inject(ElementRef);
+
+    readonly onSizeSet = output<SizeUpdate>();
+
+    ngAfterViewInit(): void {
+        this.setDimensions();
+    }
+
+    isMastered(): boolean {
+        return this.mastery() >= 80;
```
Comment (JohannesStoehr) [inline src/main/webapp/app/course/learning-paths/components/competency-node/competency-node.component.ts:35]: Shouldn't this depend on the mastery threshold?

File: src/main/webapp/app/course/learning-paths/participate/learning-path-container.component.ts [hunk 3]
```diff
@@ -206,14 +206,14 @@ export class LearningPathContainerComponent implements OnInit {
...
-            modalRef.componentInstance.learningPath = learningPathResponse.body!;
+        const modalRef = this.modalService.open(CompetencyGraphModalComponent, {
+            size: 'xl',
+            backdrop: 'static',
+            windowClass: 'competency-graph-modal',
         });
+        modalRef.componentInstance.learningPathId = this.learningPathId;
+        // modalRef.componentInstance.learningPath = learningPathResponse.body!;
+        // this.learningPathService.getLearningPath(this.learningPathId).subscribe((learningPathResponse) => {
+        // });
```
Comment (JohannesStoehr) [inline src/main/webapp/app/course/learning-paths/participate/learning-path-container.component.ts:217]: Commented code?

File: src/main/webapp/app/course/learning-paths/components/competency-node/competency-node.component.html [hunk 4]
```diff
@@ -0,0 +1,10 @@
+<div class="d-inline-block competency-node">
+    <div class="progress-container d-inline-block" [ngClass]="{ 'green-competency': isMastered(), 'yellow-competency': isStarted(), 'gray-competency': isNotStarted() }">
```
Comment (pzdr7) [inline src/main/webapp/app/course/learning-paths/components/competency-node/competency-node.component.html:2]: These are method calls, aren't they? Could we get rid of them?

File: src/test/javascript/spec/component/learning-paths/components/competency-node.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        expect(component.isMastered()).toBeFalsy();
+    });
+
+    it('should check if competency is started', () => {
+        expect(component.isStarted()).toBeTruthy();
+    });
+
+    it('should check if competency is not started', () => {
+        expect(component.isNotStarted()).toBeFalsy();
+    });
```
Comment (pzdr7) [inline src/test/javascript/spec/component/learning-paths/components/competency-node.component.spec.ts:58]: You've annotated these methods with `: boolean`. Could we instead use `toBeFalse` and `toBeTrue` here?

File: src/main/webapp/app/course/learning-paths/services/base-api-http.service.ts [hunk 6]
```diff
@@ -66,7 +66,22 @@ export abstract class BaseApiHttpService {
         return await this.request<T>(HttpMethod.Get, url, options);
     }
 
-    protected async post<T>(url: string, body?: any): Promise<T> {
-        return await this.request<T>(HttpMethod.Post, url, { body: body });
+    protected async post<T>(
```
Comment (JohannesStoehr) [inline src/main/webapp/app/course/learning-paths/services/base-api-http.service.ts:69]: Can you add some javadoc above these methods? The parameters get quite complicated...

Comment (JohannesStoehr) [inline src/main/webapp/app/course/learning-paths/services/base-api-http.service.ts:69]: I just noticed now that you use `private readonly httpClient = inject(HttpClient);` in this file. I think we use constructor injection in every other service. Can we do this here to?

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/competency/LearningPathNavigationOverviewDTO.java [hunk 7]
```diff
@@ -2,5 +2,5 @@
 
 import java.util.List;
 
-public record LearningPathNavigationOverviewDto(List<LearningPathNavigationObjectDTO> learningObjects) {
+public record LearningPathNavigationOverviewDTO(List<LearningPathNavigationObjectDTO> learningObjects) {
 }
```
Comment (pzdr7) [inline src/main/java/de/tum/in/www1/artemis/web/rest/dto/competency/LearningPathNavigationOverviewDTO.java:6]: There's a failing architecture test: testDTOImplementations - can you check if this DTO needs a JsonInclude annotation?

**Review by pzdr7 — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/course/learning-paths/services/base-api-http.service.ts (hunk 6 above)
Comment (JohannesWt) [reply src/main/webapp/app/course/learning-paths/services/base-api-http.service.ts:69]: inject is a relatively new Angular feature (came with signals) which should replace constructor injection.

File: src/main/webapp/app/course/learning-paths/components/learning-path-exercise/learning-path-exercise.component.ts [hunk 8]
```diff
@@ -8,19 +8,21 @@ import { CourseExerciseDetailsModule } from 'app/overview/exercise-details/cours
...
 
-    ngOnInit(): void {
-        this.viewContainerRef.clear();
-        // The exercise component can not be directly added to the template as before rendering the learning path mode
-        // has to be activated. This is done by setting the learningPathMode property of the exercise component to true.
-        const exerciseComponent = this.viewContainerRef.createComponent(CourseExerciseDetailsComponent);
-        exerciseComponent.instance.courseId = this.courseId();
-        exerciseComponent.instance.exerciseId = this.exerciseId();
-        exerciseComponent.instance.learningPathMode = true;
+    constructor() {
```
[the remark this answers is not in the document]
Comment (JohannesWt) [reply src/main/webapp/app/course/learning-paths/components/learning-path-exercise/learning-path-exercise.component.ts:17]: if you go to a new exercise for example only the exercise id is changing. Therefore ngOnInit is not called again. Therefore I put it in an effect in the constructor which runs every time a signal used within the effect changes.

File: src/main/webapp/app/course/learning-paths/learning-path-graph/learning-path-legend.component.html [hunk 9]
```diff
@@ -14,7 +14,7 @@
             <div class="col-auto"><fa-icon class="legend-icon" [icon]="getIcon(matchStart)" /></div>
             <div class="col-auto">
                 <span>
-                    {{ 'artemisApp.learningPath.graph.legend.matchStart.title' | artemisTranslate }}
+                    {{ 'artemisApp.learningPath.graph.legend.matchStart.label' | artemisTranslate }}
```
[the remark this answers is not in the document]
Comment (JohannesWt) [reply src/main/webapp/app/course/learning-paths/learning-path-graph/learning-path-legend.component.html:17]: Same as above (unused code)

File: src/main/webapp/app/course/learning-paths/learning-path-graph/learning-path-legend.component.html [hunk 10]
```diff
@@ -24,7 +24,7 @@
             <div class="col-auto"><fa-icon class="legend-icon" [icon]="getIcon(matchEnd)" /></div>
             <div class="col-auto">
                 <span>
-                    {{ 'artemisApp.learningPath.graph.legend.matchEnd.title' | artemisTranslate }}
+                    {{ 'artemisApp.learningPath.graph.legend.matchEnd.label' | artemisTranslate }}
```
[the remark this answers is not in the document]
Comment (JohannesWt) [reply src/main/webapp/app/course/learning-paths/learning-path-graph/learning-path-legend.component.html:27]: same unused code

File: src/main/webapp/app/course/learning-paths/participate/learning-path-container.component.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (JohannesWt) [reply src/main/webapp/app/course/learning-paths/participate/learning-path-container.component.ts:217]: see @JohannesStoehr comment (unused code - will be deleted soon)

File: src/main/webapp/app/course/learning-paths/components/competency-node/competency-node.component.html (hunk 4 above)
Comment (JohannesWt) [reply src/main/webapp/app/course/learning-paths/components/competency-node/competency-node.component.html:2]: With the use of signals only it doesn't matter anymore if you call methods within the template. More infos on that here: https://github.com/angular/angular/discussions/49682
image

**Review by rstief — CHANGES_REQUESTED** [review]
Tested on ts3. In the instructor view for learning paths translations are now missing:
image

I also get a lot of internal server errors, I'll re-test on another server and see if they are related (as i can't access grafana ATM)

**Comment by JohannesWt** [comment]
> Tested on ts3. In the instructor view for learning paths translations are now missing: image
> 
> I also get a lot of internal server errors, I'll re-test on another server and see if they are related (as i can't access grafana ATM)

This is related to TS3. Had the same issue yesterday. The screenshot also still shows the old graph. Don't know why there is this problem on TS3.

File: src/main/webapp/app/course/learning-paths/components/competency-node/competency-node.component.html (hunk 4 above)
Comment (pzdr7) [reply src/main/webapp/app/course/learning-paths/components/competency-node/competency-node.component.html:2]: The RFC you've linked also mentions:

> You can mark a component as a signal-based component by setting signals: true in the component metadata:
>
> ```
> @Component({
>   signals: true,
>   ...
> })

Can you just confirm that the component counts as signal-based even though we do not have this declaration?

**Review by rstief — COMMENTED** [review]
tested on ts1. This time I had no errors :) 
Student side:
- worked as expected.
- (Optional) In the future It would be cool if the graph was centered instead of the upper left corner.
Instructor side: Please fix the translations in the graph.

File: src/main/webapp/app/course/learning-paths/components/competency-node/competency-node.component.html (hunk 4 above)
Comment (JohannesWt) [reply src/main/webapp/app/course/learning-paths/components/competency-node/competency-node.component.html:2]: this has been removed again. Apparently Angular now self detects if only signals are used in the template.

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by rstief — APPROVED** [review]
tested on ts1 works as expected

**Review by BBesrour — APPROVED** [review]
tested on ts1, works as described

**Comment by egekurt123** [comment]
> Tested on ts3. In the instructor view for learning paths translations are now missing: image
> 
> I also get a lot of internal server errors, I'll re-test on another server and see if they are related (as i can't access grafana ATM)

This behavior also happens in ts2, but the changes work fine on TS1

**Review by egekurt123 — APPROVED** [review]

**Review by kaancayli — APPROVED** [review]
Tested on TS4. Works as expected. Code LGTM

**Review by sarpsahinalp — COMMENTED** [review]
Tested on TS4. When no tasks are completed in the learning path, the starting competency node of the graph displays an arrow that points to nothing
image

**Comment by JohannesWt** [comment]
> Tested on TS4. When no tasks are completed in the learning path, the starting competency node of the graph displays an arrow that points to nothing image

I can't reproduce it on TS4. Can you give me more insights?
image

**Pull request merged**